notebooks/03-query/11-joins.ipynb

```diff
@@ -0,0 +1,2475 @@
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Results = avg:12015.551500000001 ms   min:11389.986 ms  max:13015.264000000001 ms\n",
+      "2023-10-26 01:06:05.324785 <<<<<<<<<<<<< \n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "datetime.datetime(2023, 10, 26, 1, 6, 5, 324785)"
+      ]
+     },
+     "execution_count": 26,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "#MSQ async query using sortMerge\n",
+    "print(measure_async_query(sql,10,joinAlgorithm='sortMerge'))\n",
+    "print_timestamp()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "88a0f948-3a9e-405b-a305-95fed27db6ca",
+   "metadata": {},
+   "source": [
+    "In this case, sortMerge joins did not help which makes sense. A broadcast of 4000 `users` rows to the workers is much faster than having to redistribute both `clicks` and `users`. But if the two tables are large, a broadcast of one of them will not be possible because it would not fit in the Heap of the `query_workers`.\n",
+    "Try the following query that joins `clicks` with `clicks` in using `broadcast` and `sortMerge` algorithms to see this effect."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 27,
+   "id": "844536e5-9159-4325-be4b-adcb13974bff",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Error: Query failed: Size of broadcast tables in JOIN exceeds reserved memory limit (memory reserved for broadcast tables = 26541947 bytes). Increase available memory, or set sqlJoinAlgorithm: sortMerge in query context to use a shuffle-based join.\n"
+     ]
+    }
+   ],
+   "source": [
+    "sql = '''\n",
+    "SELECT c.__time, c.user_id, c.client_ip, MAX(c2.__time) prior_time_in_session, count(*) prior_events_in_session \n",
+    "FROM clicks c\n",
+    "     INNER JOIN \n",
+    "      clicks c2 ON c.user_id=c2.user_id AND c.client_ip = c2.client_ip\n",
+    "WHERE c.__time > (CURRENT_TIMESTAMP  - INTERVAL '1' DAY) AND\n",
+    "      c2.__time > (CURRENT_TIMESTAMP - INTERVAL '2' DAY) AND\n",
+    "      c2.__time < c.__time\n",
+    "GROUP BY 1,2,3\n",
+    "LIMIT 10\n",
+    "'''\n",
+    "try:\n",
+    "    result = sql_client.async_sql(sql)\n",
+    "    result.rows\n",
+    "except Exception as ex:\n",
+    "    print(f\"Error: {ex}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d6cded7c-2019-40d7-a7dd-0ebeb5d5a134",
+   "metadata": {},
+   "source": [
+    "As expected, the query fails because it cannot broadcast that much data to the workers. As the suggestion in the error indicates, the next run of the same query uses `sortMerge` for the join algorithm which does not have such limits. \n",
+    "\n",
+    "It takes a couple of minutes to complete, while it is running you can see its progress by attaching to it in the [Druid Console](http://localhost:8888/unified-console.html#workbench):\n",
```

Comment:
Is there a way to show this without using the console? An API call?

Reply:
API output is hard to read.